File: 2015/day2.hs
#!/usr/bin/env runhaskell
module Main where
import Data.List

-- draws from & is inspired by travisby's implementation
-- (https://gist.github.com/travisby/0fa62b22f860afa01a93)

main = do
    contents <- getContents
    let packages = map (strToList) $ lines contents
    print $ total packages

total :: [[Int]] -> Int
total xs = sum $ map single xs

single :: [Int] -> Int
single xs = 2 * (sum ys) + (minimum ys)
    where ys = areas xs

areas :: [Int] -> [Int]
areas xs = map (\(x,y) -> x*y) (zip xs $ rotate xs)

rotate :: [a] -> [a]
rotate xs = drop 1 xs ++ take 1 xs

strToList :: String -> [Int]
strToList str = map read $ split str 'x'

split :: String -> Char -> [String]
split xs c = foldr (splitOne c) [] xs

splitOne :: Char -> Char -> [String] -> [String]
splitOne _ x [] = [[x]]
splitOne a b (x:xs)
    | a == b = []:x:xs
    | otherwise = (b:x):xs

File: 2015/day2.rb
#!/usr/bin/env ruby

total = 0
total_ribbon = 0

STDIN.each_line do |line|
  line.chomp!
  sizes = line.split('x').map(&:to_i)
  areas = sizes.zip(sizes.rotate).map { |a| a.reduce :* }
  wrapping = areas.reduce(&:+)*2 + areas.min
  sizes = sizes.sort
  ribbon = 2*(sizes[0] + sizes[1]) + sizes.reduce(:*)

  total += wrapping
  total_ribbon += ribbon
end

puts "Wrapping paper: #{total}"
puts "Ribbon: #{total_ribbon}"
